### PR TITLE
[RW-1229][Risk=no]Adding Concept Sets to RecentResource Cache

### PR DESCRIPTION
**Pull request by UrsaStutsman**
Should be done, just trying to keep up with the db changelog and merge conflicts

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java [hunk 1]
```diff
@@ -84,6 +88,23 @@ public void updateCohortEntry(long workspaceId, long userId, long cohortId,
...
+                                   Timestamp lastAccessDateTime) {
+    ConceptSet conceptSet = conceptSetDao.findOne(conceptSetId);
+    UserRecentResource resource = getDao().findByUserIdAndWorkspaceIdAndConceptSet(userId, workspaceId, conceptSet);
+    if (resource == null) {
+      handleUserLimit(userId);
+      resource = new UserRecentResource();
+      resource.setUserId(userId);
+      resource.setWorkspaceId(workspaceId);
+      resource.setConceptSet(conceptSet);
+      resource.setNotebookName(null);
```
Comment (rushtong) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: Not a request - just a comment for posterity. I wish our dao objects had a builder pattern constructor.

File: api/db/changelog/db.changelog-48-add-concept-set-to-user-recent-resource.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+                <constraints nullable="true"/>
+            </column>
+        </addColumn>
+        <addForeignKeyConstraint baseTableName="user_recent_resource"
+                                 baseColumnNames="concept_set_id"
+                                 constraintName="fk_cache_concept_set_id"
+                                 referencedTableName="concept_set"
+                                 referencedColumnNames="concept_set_id" onDelete="CASCADE"/>
+    </changeSet>
+</databaseChangeLog>
```
Comment (rushtong) [inline api/db/changelog/db.changelog-48-add-concept-set-to-user-recent-resource.xml]: nit: newline

File: api/src/main/java/org/pmiops/workbench/db/model/UserRecentResource.java [hunk 3]
```diff
@@ -63,5 +64,28 @@ public void setLastAccessDate(Timestamp lastAccessDate) {
...
+    this.conceptSet = null;
+  }
+
+  public UserRecentResource(long workspaceId, long userId, Timestamp lastAccessDate) {
+    this.workspaceId = workspaceId;
+    this.userId = userId;
+    this.notebookName = null;
+    this.lastAccessDate = lastAccessDate;
+  }
+}
```
Comment (rushtong) [inline api/src/main/java/org/pmiops/workbench/db/model/UserRecentResource.java]: Thank you!

File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java [hunk 4]
```diff
@@ -50,11 +56,7 @@ public UserRecentResource updateNotebookEntry(long workspaceId, long userId, Str
     UserRecentResource recentResource = getDao().findByUserIdAndWorkspaceIdAndNotebookName(userId, workspaceId, notebookNameWithPath);
     if (recentResource == null) {
       handleUserLimit(userId);
-      recentResource = new UserRecentResource();
-      recentResource.setUserId(userId);
-      recentResource.setWorkspaceId(workspaceId);
-      recentResource.setCohort(null);
-      recentResource.setNotebookName(notebookNameWithPath);
+      recentResource = new UserRecentResource(workspaceId, userId, notebookNameWithPath, lastAccessDateTime);
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: I think you could also use the builder pattern here and below (Greg just taught me this one!)
```
recentResource = new UserRecentResource()
    .workspaceId(workspaceId)
    .userId(userId)
    .notebookName(notebookNameWithPath)
    .lastAccessDate(lastAccessDateTime)
```
Or something? And that would avoid the multiple defined constructors.

Comment (rushtong) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: Sorry @blrubenstein - our entities are not built that way :-(  Model classes derived from swagger get that pattern for free, so wherever you use one of those, you can use that pattern.

Comment (blrubenstein) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: Oh no... okay... :(

Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: FYI it also makes me sad that they aren't symmetric as I find this pattern much more readable. I filed https://precisionmedicineinitiative.atlassian.net/browse/RW-1384 a few weeks ago to apply this pattern throughout our models (Cohort Builder models currently do have chaining, actually).

**Pull request merged**